pkg/controller/volume/persistentvolume/topology_binder.go

```diff
@@ -0,0 +1,482 @@
+}
+
+// FindPodVolumes internally caches the matching PVs per node in tmpData.nodeBindings
+func (b *topologyVolumeBinder) FindPodVolumes(pod *v1.Pod, nodeName string) (needsBinding bool, foundMatches bool, err error) {
+	glog.V(4).Infof("FindVolumesToBind for pod %q, node %q", getPodName(pod), nodeName)
+
+	node, err := b.nodeCache.Get(nodeName)
+	if node == nil || err != nil {
+		return false, false, fmt.Errorf("error getting node %q: %v", nodeName, err)
+	}
+
+	boundClaims, unboundClaims, unboundClaimsImmediate, err := b.getPodVolumes(pod)
+	if err != nil {
+		return false, false, err
+	}
+
+	// Immediate claims should be bound
+	if len(unboundClaimsImmediate) > 0 {
+		return true, false, fmt.Errorf("pod has unbound PersistentVolumeClaims")
+	}
+
+	// Check PV node affinity on bound volumes
+	if len(boundClaims) > 0 {
+		affinityMatches, err := b.checkBoundClaims(boundClaims, node, getPodName(pod))
+		if err != nil || !affinityMatches {
+			return false, affinityMatches, err
+		}
+	}
+
+	// Find PVs for unbound volumes
+	if len(unboundClaims) > 0 {
+		foundMatches, err := b.findMatchingVolumes(unboundClaims, node)
+		return true, foundMatches, err
+	}
+
+	// Pod has no PVC volumes, return success
+	return false, true, nil
+}
+
+func (b *topologyVolumeBinder) checkBoundClaims(claims []*v1.PersistentVolumeClaim, node *v1.Node, podName string) (bool, error) {
+	for _, pvc := range claims {
+		pvName := pvc.Spec.VolumeName
+		pv := b.pvCache.GetPV(pvName)
+		if pv == nil {
+			return false, fmt.Errorf("PersistentVolume %q not found", pvName)
+		}
+
+		err := volumeutil.CheckNodeAffinity(pv, node.Labels)
+		if err != nil {
+			glog.V(4).Infof("PersistentVolume %q, Node %q mismatch for Pod %q: %v", pvName, node.Name, err.Error(), podName)
+			return false, nil
+		}
+		glog.V(5).Infof("PersistentVolume %q, Node %q matches for Pod %q", pvName, node.Name, podName)
+	}
+
+	glog.V(4).Infof("All volumes for Pod %q match with Node %q", podName, node.Name)
+	return true, nil
+}
+
+func (b *topologyVolumeBinder) findMatchingVolumes(claimsToBind []*bindingInfo, node *v1.Node) (foundMatches bool, err error) {
+	// Sort all the claims by increasing size request to get the smallest fits
+	sort.Sort(byPVCSize(claimsToBind))
+
+	allPVs := b.pvCache.ListPVs()
+	chosenPVs := map[string]*v1.PersistentVolume{}
+
+	for _, bindingInfo := range claimsToBind {
+		// Find a matching PV
+		bindingInfo.pv, err = findMatchingVolume(bindingInfo.pvc, allPVs, node, chosenPVs, true)
+		if err != nil {
+			return false, err
+		}
+		if bindingInfo.pv == nil {
+			glog.V(4).Infof("No matching volumes for PVC %q on node %q", getPVCName(bindingInfo.pvc), node.Name)
+			return false, nil
+		}
+
+		// matching PV needs to be excluded so we don't select it again
+		chosenPVs[bindingInfo.pv.Name] = bindingInfo.pv
+	}
+
+	// Mark cache with all the matches for each PVC for this node
+	b.tmpData.nodeBindings[node.Name] = claimsToBind
```

Comment:
This needs to be protected by a mutex because FindPodVolumes() can be called in parallel across nodes